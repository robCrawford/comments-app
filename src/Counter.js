/*eslint-disable no-unused-vars */
import React from "react";
import { Component, PropTypes } from 'prop-types';

const Counter = ({ value, onIncrement, onDecrement, onIncrementAsync }) =>
  <div>
    <button onClick={onIncrementAsync}>
      Increment async
    </button>
    {' '}
    <button onClick={onIncrement}>
      Increment
    </button>
    {' '}
    <button onClick={onDecrement}>
      Decrement
    </button>
    <hr />
    <div>
      Counter: {value}
    </div>
  </div>;

Counter.propTypes = {
  value: PropTypes.number.isRequired,
  onIncrement: PropTypes.func.isRequired,
  onDecrement: PropTypes.func.isRequired,
  onIncrementAsync: PropTypes.func.isRequired
};

export default Counter;
