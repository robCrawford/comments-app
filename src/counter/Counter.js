/*eslint-disable no-unused-vars */
import React from "react";
import { useSelector } from "react-redux";
import { appSelector } from "../app/app-redux";
import { counterSelector } from './counter-redux';

const Counter = ({ onIncrement, onDecrement, onSync }) => {
  const { count } = useSelector(counterSelector);
  const { loading } = useSelector(appSelector);
  const handleIncrement = () => !loading && onIncrement();
  const handleDecrement = () => !loading && onDecrement();
  const handleSync = () => !loading && onSync();

  return (<div>
    <button onClick={handleIncrement}>
      Increment
    </button>
    <button onClick={handleDecrement}>
      Decrement
    </button>
    <button onClick={handleSync}>
      { loading ? 'Loading...' : 'Sync' }
    </button>
    <p>
      Counter: {count}
    </p>
  </div>);
};

export default Counter;
