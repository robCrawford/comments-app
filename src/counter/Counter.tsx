/*eslint-disable no-unused-vars */
import React, { FC } from 'react';
import { useSelector } from 'react-redux';
import { appSelector } from '../app/app-redux';
import { counterSelector } from './counter-redux';

type Props = {
  onIncrement: () => void;
  onDecrement: () => void;
  onSync: () => void;
};

const Counter: FC<Props> = ({ onIncrement, onDecrement, onSync }) => {
  const { count } = useSelector(counterSelector);
  const { loading } = useSelector(appSelector);
  const handleIncrement = (): void => {
    !loading && onIncrement();
  };
  const handleDecrement = (): void => {
    !loading && onDecrement();
  };
  const handleSync = (): void => {
    !loading && onSync();
  };

  return (<div>
    <button onClick={handleIncrement}>
      Increment
    </button>
    <button onClick={handleDecrement}>
      Decrement
    </button>
    <button onClick={handleSync}>
      { loading ? 'Loading...' : 'Sync' }
    </button>
    <p>
      Counter: {count}
    </p>
  </div>);
};

export default Counter;
