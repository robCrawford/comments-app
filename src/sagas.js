import { all, call, put, takeEvery } from 'redux-saga/effects';

export const apiCall = () => new Promise(res => setTimeout(() => res(Math.floor(Math.random() * 10) + 100), 1500));

function* init() {
  console.log('init');
  yield undefined;
}

// Worker Saga
export function* setFromApi() {
  yield put({ type: 'LOADING' });
  const result = yield call(apiCall);
  yield put({ type: 'SET', 'payload': result });
  yield put({ type: 'CLEAR_LOADING' });
}

// Watcher Saga
function* watchSetFromApi() {
  yield takeEvery('SET_FROM_API', setFromApi);
}

export default function* rootSaga() {
  yield all([
    init(),
    watchSetFromApi()
  ]);
}
