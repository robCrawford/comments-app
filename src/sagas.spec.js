// import test from 'tape';


// https://dev.to/phil/the-best-way-to-test-redux-sagas-4hib
// import {runSaga} from 'redux-saga';

// async function recordSaga(saga, initialAction) {
//   const dispatched = [];

//   await runSaga(
//     {
//       dispatch: (action) => dispatched.push(action)
//     },
//     saga,
//     initialAction
//   ).done;

//   return dispatched;
// }



/* import { put, call } from 'redux-saga/effects';
import { incrementAsync, delay } from './sagas';

test('incrementAsync Saga test', (assert) => {
  const gen = incrementAsync();

  assert.deepEqual(
    gen.next().value,
    call(delay, 1000),
    'incrementAsync Saga must call delay(1000)'
  );

  assert.deepEqual(
    gen.next().value,
    put({type: 'INCREMENT'}),
    'incrementAsync Saga must dispatch an INCREMENT action'
  );

  assert.deepEqual(
    gen.next(),
    { done: true, value: undefined },
    'incrementAsync Saga must be done'
  );

  assert.end();
}); */

import React from 'react';
import { shallow } from 'enzyme';
import Counter from './Counter';

describe('Counter', () => {
  it('should render', () => {
    const component = shallow(<Counter
      value={0}
      onIncrement={() => {}}
      onDecrement={() => {}}
      onIncrementAsync={() => {}} />);

    expect(component).toMatchSnapshot();
  });
});

