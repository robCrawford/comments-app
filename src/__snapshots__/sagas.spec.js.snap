// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should render 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Increment async
  </button>
   
  <button
    onClick={[Function]}
  >
    Increment
  </button>
   
  <button
    onClick={[Function]}
  >
    Decrement
  </button>
  <hr />
  <div>
    Counter: 
    0
  </div>
</div>
`;
